Description: Allow compilation on kFreeBSD.
From: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=640378
Forwarded: http://code.google.com/p/leveldb/issues/detail?id=38

Index: bitcoin/src/leveldb/build_detect_platform
===================================================================
--- bitcoin.orig/src/leveldb/build_detect_platform	2013-09-03 23:02:09.861470907 -0400
+++ bitcoin/src/leveldb/build_detect_platform	2013-09-03 23:02:09.853470907 -0400
@@ -145,6 +145,12 @@
         PORT_FILE=port/port_win.cc
         CROSS_COMPILE=true
         ;;
+    GNU/kFreeBSD)
+        PLATFORM=OS_FREEBSD
+        PLATFORM_CFLAGS="-pthread -DOS_FREEBSD"
+        PLATFORM_LIBS="-lpthread -lrt"
+        PORT_FILE=port/port_posix.cc
+        ;;
     *)
         echo "Unknown platform!" >&2
         exit 1
