Description: Add debug output trying to pinpoint bug #672524
Author: Petter Reinholdtsen <pere@hungry.com>
Last-Update: 2013-02-02

Index: bitcoin/src/test/script_tests.cpp
===================================================================
--- bitcoin.orig/src/test/script_tests.cpp	2013-03-20 23:04:15.760259263 -0400
+++ bitcoin/src/test/script_tests.cpp	2013-03-20 23:04:15.756259263 -0400
@@ -98,6 +98,9 @@
     }
 #endif
 
+    // Debug #672524
+    cerr << "Trying to open " << testFile << endl;
+
     ifstream ifs(testFile.string().c_str(), ifstream::in);
     Value v;
     if (!read_stream(ifs, v))
