Description: Add debug output trying to pinpoint bug #672524
Author: Petter Reinholdtsen <pere@hungry.com>
Last-Update: 2013-02-02

diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index 61d9a64..8762364 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -98,6 +98,9 @@ read_json(const std::string& filename)
     }
 #endif
 
+    // Debug #672524
+    cerr << "Trying to open " << testFile << endl;
+
     ifstream ifs(testFile.string().c_str(), ifstream::in);
     Value v;
     if (!read_stream(ifs, v))
